src/copc.rs

```diff
@@ -0,0 +1,395 @@
+    fn read_from<R: Read>(read: &mut R) -> Result<Self> {
+        Ok(Self {
+            key: VoxelKey::read_from(read)?,
+            offset: read.read_u64::<LittleEndian>()?,
+            byte_size: read.read_i32::<LittleEndian>()?,
+            point_count: read.read_i32::<LittleEndian>()?,
+        })
+    }
+    fn write_to<W: Write>(&self, dst: &mut W) -> Result<()> {
+        self.key.write_to(dst)?;
+        dst.write_u64::<LittleEndian>(self.offset)?;
+        dst.write_i32::<LittleEndian>(self.byte_size)?;
+        dst.write_i32::<LittleEndian>(self.point_count)?;
+        Ok(())
+    }
+    fn is_referencing_page(&self) -> bool {
+        self.point_count == -1
+    }
+}
+
+/// The entries of a hierarchy page are consecutive. The number of entries in a page
+/// can be determined by taking the size of the page (contained in the parent page as
+/// Entry::byteSize or in the COPC info VLR as CopcData::root_hier_size)
+/// and dividing by the size of an Entry (32 bytes).
+#[derive(Debug)]
+struct Page {
+    entries: Vec<Entry>, //[page_size / 32];
+}
+impl Page {
+    /// Reads hierarchy page from a `Read`.
+    fn read_from(mut data: &[u8]) -> Result<Self> {
+        Ok(Self {
+            entries: (0..data.len() / 32)
+                .map(|_| Entry::read_from(&mut data))
+                .collect::<Result<Vec<Entry>>>()?,
+        })
+    }
+    fn write_to<W: Write>(&self, dst: &mut W) -> Result<()> {
+        self.entries
+            .iter()
+            .try_for_each(|entry| entry.write_to(dst))?;
+        Ok(())
+    }
+}
+
+/// the hierarchy VLR MUST exist.
+/// Like EPT, COPC stores hierarchy information to allow a reader to locate points
+/// that are in a particular octree node. Also like EPT, the hierarchy MAY be
+/// arranged in a tree of pages, but SHALL always consist of at least ONE hierarchy
+/// page.
+/// The VLR data consists of one or more hierarchy pages. Each hierarchy data
+/// page is written as follows:
+/// The VoxelKey corresponds to the naming of EPT data files.
+///
+/// The octree hierarchy is arranged in pages. The COPC VLR provides information
+/// describing the location and size of root hierarchy page. The root hierarchy page
+/// can be used to traverse to child pages. Each entry in a hierarchy page either
+/// refers to a child hierarchy page, octree node data chunk, or an empty octree
+/// node. The size and file offset of each data chunk is provided in the hierarchy
+/// entries, allowing the chunks to be directly read for decoding.
+#[derive(Debug)]
+pub struct CopcHierarchyVlr {
+    root: Page,
+    sub_pages: HashMap<VoxelKey, Page>,
+}
+impl CopcHierarchyVlr {
+    /// The user id of the LasZip VLR header.
+    pub const USER_ID: &'static str = "copc";
+    /// The record id of the LasZip VLR header.
+    pub const RECORD_ID: u16 = 1000;
+    /// The description of the LasZip VLR header.
+    pub const DESCRIPTION: &'static str = "https://copc.io";
```

Comment:
The user id and description are duplicated between the two VLR types, so let's make them module-level constants.